ZCS-14487 - New install script to support installs with new license server and all supported OS versions.

GaneshAnarse opened the pull request:
**ZCS-14487 - New install script to support installs with new license server and all supported OS versions.**

- New script is an attempt to create a new unified auto install script which can be used across test/dev platforms.
- Its a wrapper script around manual install process of installing Zimbra. (`supported-os-installer.sh`)
- This script should support installing all version of zimbra for all zimbra supported OS versions. 
- Tested on ubuntu20, ubuntu18, rhel7, rhel8 OS versions.
- This script should also support installation of new 10.1.0 version which requires license key instead of license.xml.

**Testing**
- Currently, this script can be used to install zimbra on a SINGLE NODE ONLY.
- License xml or License key is required argument depending on which ZCS version you are trying to install.
- You can provide both License key or License xml file path with `--license` option
- Run following for help -
```
supported-os-installer.sh --help
```
- For 10.0.0 / 9.0.0 / 8.8.15 versions - 
```
supported-os-installer.sh -l /path/to/license.xml --optional-pkg <pkg1> --optional-pkg <pkg2>
```
- For versions 10.1x.xx
```
supported-os-installer.sh -l <license-key> --optional-pkg <pkg1> --optional-pkg <pkg2>
```
- The file `zim-install-config` should serve as a default install config file. You can also modify it according to your own needs. But make sure not to replace any string that start with `placeholder_xxxx`. These get replaced automatically when the script runs. Please review it and suggest anymore settings you believe are important.

**Review Notes** 
- Please be aware that in order to use a single install script and not maintain many versions of same script to do the install, we will be deprecating / deleting some of the older scripts present on this repository.

- Please suggest more features / improvements which you believe we need in this install script. 
   They will be added as we further improve the script.

Commit:
ZCS-14487: Install script to support all OS and new license activation workflow.

Commit:
ZCS-14487: Fixed bugs related to dnsmasq, install-config. Fixed Indentions.

Commit:
ZCS-14487: Install zimbra-license-daemon only for versions >= 10.1.0

silentsakky commented on supported-os-installer.sh:
```diff
@@ -0,0 +1,524 @@
+# This script is mostly based on ->
+# https://github.com/Zimbra/zinstaller/blob/main/zinstaller script
+
+# Warnings :	This script is still a WORK IN PROGRESS. Not fully tested.
+#		Script to install Zimbra Classic - ONLY SINGLE NODE on a provided remote server
+############################################################################################
```
should we update this now?

GaneshAnarse replied:
Yes, Removed comments which are not required now!

silentsakky commented on supported-os-installer.sh:
```diff
@@ -0,0 +1,524 @@
+	options:
+	-h, --help			Help
+	-p, --admin-pass		Admin account password to set. (Default: Random)
+	-l, --license			License key or License file to be activated on the zimbra server. (Required) (No default)
```
I think barry's script supports FOSS install also, so we should also support it

GaneshAnarse replied:
Hi @silentsakky sir, 
The `zinstaller` script did not take license as an argument.
Instead it used a free trial license to install Network version of zimbra.
The script used to activate the trial license and setup the zimbra installation.

One thing we can do here is to make the `--license` arg optional.
The script will check if the license is given -> If not then it will check wether its a FOSS installation -> If yes then continue else exit!

Please let me know what do you think about this approach.

barrydegraaff replied:
FOSS was not supported in original script, which was mostly written by Gautam, not me. 

A trial license was automatically obtained and installed. This needs to be re-implemented, but the current license server does not support it.

silentsakky replied:
ok then we will go ahead with current implementation and then in future can revisit and add this support as well

silentsakky commented on supported-os-installer.sh:
```diff
@@ -0,0 +1,524 @@
+
+	[[ ! -f "$BUILD_ARTIFACT" ]] && echo "Cannot locate the build artifact (zcs-*.tgz) file." && exit 1;
+	[[ -z "$HOSTNAME" ]] && echo "Cannot find \$HOSTNAME." && exit 1;
+	[[ -z "$ADMIN_PASS " ]] && echo "Admin password is empty. Using default password." && ADMIN_PASS="zimbra"
```
should be random password

GaneshAnarse replied:
True, Modified script to Generate random 12 characters password for admin if it is not provided as argument!

silentsakky commented on supported-os-installer.sh:
```diff
@@ -0,0 +1,524 @@
+	echo "Please check your iptables for more info."
+
+	# Array of all ports required.
+	ports_arr=("25" "80" "110" "143" "443" "465" "587" "993" "995" "7071" "9071")
```
not sure if we should open 7071 port

GaneshAnarse replied:
Yes, 9071 is already being opened!
Not opening the port 7071 in the script!

silentsakky commented on supported-os-installer.sh:
```diff
@@ -0,0 +1,517 @@
+	cat << here
+	Your new Zimbra installation details are:
+	- Webmail Login:	https://${HOSTNAME}
+	- Admin Console:	https://${HOSTNAME}:7071
```
```suggestion
	- Admin Console:	https://${HOSTNAME}:9071
```

GaneshAnarse replied:
Using 9071 for reporting admin panel location now!

ghost commented:
I am trying to use the new script, and while it may work, it is not user friendly.

1. Begins by the use of `#!/usr/bin/env bash` this will fill the screen with debug information, obscuring whatever output is relevant to the user. Suggest to use: #!/bin/bash
2. **Next it requires a `build-artifact` please use plain English, you could say something like Zimbra binary in tgz format. But this should automatically download and not bother the user, at least it should have the option of getting all GA releases based on the operating system the script is running on.**
3. This script also needs a config file, it would have been nice if that was also auto downloaded and or self-contained.
4. Please update the README.md

At least item 2 needs to be address if it is to replace zinstaller.

GaneshAnarse commented:
> I am trying to use the new script, and while it may work, it is not user friendly.
> 
> 1. Begins by the use of `#!/usr/bin/env bash` this will fill the screen with debug information, obscuring whatever output is relevant to the user. Suggest to use: #!/bin/bash
> 2. **Next it requires a `build-artifact` please use plain English, you could say something like Zimbra binary in tgz format. But this should automatically download and not bother the user, at least it should have the option of getting all GA releases based on the operating system the script is running on.**
> 3. This script also needs a config file, it would have been nice if that was also auto downloaded and or self-contained.
> 4. Please update the README.md
> 
> At least item 2 needs to be address if it is to replace zinstaller.

Thank you for the pointers in improving the script. I’ll make a sure to address all the points you mentioned.

Commit:
Improvement: Download default 10.0.0_GA build to install.
* Renamed variable: BUILD_ARTIFACT -> ZCS_BINARY_TGZ
* Using /bin/bash shell for execution.

Commit:
Improvement: Update the README.md

GaneshAnarse commented:
@barrydegraaff ,
Apologies for delay in addressing the modifications you posted.
I have address and modified following points - 
- Using `#!/bin/bash` instead of `#!/usr/bin/env bash`.
- Renamed the variable: `BUILD_ARTIFACT` to `ZCS_BINARY_TGZ`
- The script will now donwload the `10.0.0_GA` build based on the OS if the zcs binary path is not provided.
- Updated the README.md.

About the config file - If you are talking about the install config file which is used for automated installs of Zimbra - I have added `zim-install-config` file already. I am not sure if this script needs any other config file. Please suggest.

Could you please review the PR again. Thanks!
